File: backend/app/detection/narrative_tracker.py
"""Narrative State Machine for scam call phase tracking.

Tracks the temporal progression of a scam call through manipulation phases:
  IDLE → HOOK → ESCALATE → ISOLATE → TRAP

Each archetype defines which keyword categories trigger each phase transition.
The tracker consumes transcript chunks over time and advances the state when
the expected phase's trigger keywords appear.

Key insight: scam calls follow a predictable psychological manipulation arc.
Detecting the *sequence* of manipulation tactics — not just individual keywords —
dramatically reduces false positives and enables early warning.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field

from app.detection.scam_archetypes import (
    NARRATIVE_PHASES,
    PHASE_ESCALATE,
    PHASE_HOOK,
    PHASE_IDLE,
    PHASE_ISOLATE,
    PHASE_TRAP,
    SCAM_ARCHETYPES,
    _match_keywords,
)


def _match_keywords_strict(text_lower: str, words: set[str], keywords: set[str]) -> str | None:
    """Stricter keyword matching — substring only, no stem/prefix matching.

    Used for phase-skip checks where loose matching causes false positives
    (e.g. 'card' in 'your new card will be sent' matching 'card details').
    """
    for kw in keywords:
        kw_lower = kw.lower()
        if kw_lower in text_lower:
            return kw
    return None

logger = logging.getLogger(__name__)

PHASE_ORDER = [PHASE_IDLE, PHASE_HOOK, PHASE_ESCALATE, PHASE_ISOLATE, PHASE_TRAP]
PHASE_SCORES = {
    PHASE_IDLE: 0,
    PHASE_HOOK: 15,
    PHASE_ESCALATE: 40,
    PHASE_ISOLATE: 65,
    PHASE_TRAP: 95,
}


@dataclass
class NarrativeState:
    """Current state of narrative tracking for a single archetype."""
    archetype_id: str
    phase: str = PHASE_IDLE
    phase_index: int = 0
    confidence: float = 0.0
    history: list[str] = field(default_factory=list)
    trigger_log: list[str] = field(default_factory=list)


@dataclass
class NarrativeResult:
    """Result of narrative analysis across all archetypes for a session."""
    best_archetype: str | None = None
    best_phase: str = PHASE_IDLE
    phase_score: int = 0
    confidence: float = 0.0
    phase_label: str = ""
    all_states: dict[str, NarrativeState] = field(default_factory=dict)


class NarrativeTracker:
    """Tracks scam narrative progression across transcript chunks.

    Usage:
        tracker = NarrativeTracker()
        for chunk in transcript_chunks:
            result = tracker.advance(chunk)
            print(result.best_phase, result.phase_score)
    """

    def __init__(self) -> None:
        self._states: dict[str, NarrativeState] = {}
        self._archetype_map: dict[str, tuple[set[str], set[str], set[str]]] = {}

        # Build lookup: archetype_id -> (context, threat, demand) keyword sets
        for label, ctx, thr, dem in SCAM_ARCHETYPES:
            self._archetype_map[label] = (ctx, thr, dem)
            if label in NARRATIVE_PHASES:
                self._states[label] = NarrativeState(archetype_id=label)

    def advance(self, text: str) -> NarrativeResult:
        """Process a transcript chunk and advance narrative states.

        Each archetype's state machine checks whether the current chunk
        contains keywords matching the *next expected phase's* trigger type.
        If so, the state advances.  If the next phase doesn't match but a
        *later* phase does (scammers skip steps), we advance to that phase,
        reducing confidence slightly for the skip.

        Returns the best (furthest-advanced) archetype result.
        """
        if not text or not text.strip():
            return self._build_result()

        lower = text.lower()
        words = set(lower.split())

        for arch_id, state in self._states.items():
            if arch_id not in NARRATIVE_PHASES:
                continue
            if arch_id not in self._archetype_map:
                continue

            phases_def = NARRATIVE_PHASES[arch_id]
            phase_list = phases_def["phases"]
            ctx_kws, thr_kws, dem_kws = self._archetype_map[arch_id]

            next_idx = state.phase_index

            if next_idx >= len(phase_list):
                continue  # Already at TRAP, fully advanced

            # Try to advance through as many phases as this chunk supports.
            # A single chunk like "press 1 to avoid disconnection" may
            # contain both urgency (ISOLATE) and demand (TRAP) triggers.
            advanced_this_chunk = False
            while next_idx < len(phase_list):
                next_phase_def = phase_list[next_idx]
                trigger_type = next_phase_def["triggers"]

                matched = self._check_trigger(
                    lower, words, trigger_type, arch_id,
                    ctx_kws, thr_kws, dem_kws, phases_def
                )

                if matched:
                    state.phase = next_phase_def["id"]
                    state.phase_index = next_idx + 1
                    state.confidence = min(1.0, state.confidence + 0.25)
                    state.history.append(state.phase)
                    state.trigger_log.append(
                        f"{state.phase}: {trigger_type}={matched}"
                    )
                    logger.info(
                        "NARRATIVE | %s advanced to %s (trigger=%s, matched=%s)",
                        arch_id, state.phase, trigger_type, matched
                    )
                    advanced_this_chunk = True
                    next_idx = state.phase_index  # Check next phase too
                else:
                    break  # This phase didn't match; stop sequential advance

            if not advanced_this_chunk:
                # Phase skip: check if any later phase matches this chunk.
                # Real scammers don't always follow the exact sequence —
                # they may jump from ESCALATE directly to TRAP.
                # GUARD: Only allow skipping if HOOK has already been
                # reached (phase_index >= 1).  Skipping from IDLE would
                # cause false positives on innocent conversations.
                if state.phase_index < 1:
                    continue  # Must reach HOOK first before any skipping

                for skip_idx in range(next_idx + 1, len(phase_list)):
                    skip_phase_def = phase_list[skip_idx]
                    skip_trigger = skip_phase_def["triggers"]
                    # Use STRICT matching for skips — no stem/prefix
                    # matching, only exact substring. This prevents
                    # 'card' in 'new card sent' from matching 'card details'.
                    skip_matched = self._check_trigger_strict(
                        lower, words, skip_trigger, arch_id,
                        ctx_kws, thr_kws, dem_kws, phases_def
                    )
                    if skip_matched:
                        skipped = skip_idx - next_idx
                        state.phase = skip_phase_def["id"]
                        state.phase_index = skip_idx + 1
                        # Reduce confidence slightly for skipped phases
                        state.confidence = min(
                            1.0, state.confidence + 0.25 - (0.05 * skipped)
                        )
                        state.history.append(state.phase)
                        state.trigger_log.append(
                            f"{state.phase}: {skip_trigger}={skip_matched} (skipped {skipped})"
                        )
                        logger.info(
                            "NARRATIVE | %s SKIPPED to %s (trigger=%s, matched=%s, skipped=%d)",
                            arch_id, state.phase, skip_trigger, skip_matched, skipped
                        )
                        break

        return self._build_result()

    def _check_trigger(
        self,
        lower: str,
        words: set[str],
        trigger_type: str,
        arch_id: str,
        ctx_kws: set[str],
        thr_kws: set[str],
        dem_kws: set[str],
        phases_def: dict,
    ) -> str | None:
        """Check if text contains keywords for the given trigger type."""
        if trigger_type == "context":
            return _match_keywords(lower, words, ctx_kws)
        elif trigger_type == "threat":
            return _match_keywords(lower, words, thr_kws)
        elif trigger_type == "demand":
            return _match_keywords(lower, words, dem_kws)
        elif trigger_type == "secrecy":
            extra = phases_def.get("secrecy_keywords", set())
            return _match_keywords(lower, words, extra)
        elif trigger_type == "urgency":
            extra = phases_def.get("urgency_keywords", set())
            return _match_keywords(lower, words, extra)
        elif trigger_type == "authority":
            extra = phases_def.get("authority_keywords", set())
            return _match_keywords(lower, words, extra)
        elif trigger_type == "social_proof":
            extra = phases_def.get("social_proof_keywords", set())
            return _match_keywords(lower, words, extra)
        elif trigger_type == "control":
            extra = phases_def.get("control_keywords", set())
            return _match_keywords(lower, words, extra)
        return None

    def _check_trigger_strict(
        self,
        lower: str,
        words: set[str],
        trigger_type: str,
        arch_id: str,
        ctx_kws: set[str],
        thr_kws: set[str],
        dem_kws: set[str],
        phases_def: dict,
    ) -> str | None:
        """Strict trigger check for phase skipping — substring only, no stem matching."""
        if trigger_type == "context":
            return _match_keywords_strict(lower, words, ctx_kws)
        elif trigger_type == "threat":
            return _match_keywords_strict(lower, words, thr_kws)
        elif trigger_type == "demand":
            return _match_keywords_strict(lower, words, dem_kws)
        else:
            extra_key = f"{trigger_type}_keywords"
            extra = phases_def.get(extra_key, set())
            return _match_keywords_strict(lower, words, extra)

    def _build_result(self) -> NarrativeResult:
        """Build result from the archetype with the most advanced phase."""
        best_arch = None
        best_phase = PHASE_IDLE
        best_idx = 0
        best_conf = 0.0

        for arch_id, state in self._states.items():
            if state.phase_index > best_idx or (
                state.phase_index == best_idx and state.confidence > best_conf
            ):
                best_arch = arch_id
                best_phase = state.phase
                best_idx = state.phase_index
                best_conf = state.confidence

        phase_score = PHASE_SCORES.get(best_phase, 0)

        phase_label = ""
        if best_arch and best_arch in NARRATIVE_PHASES:
            for p in NARRATIVE_PHASES[best_arch]["phases"]:
                if p["id"] == best_phase:
                    phase_label = p["description"]
                    break

        return NarrativeResult(
            best_archetype=best_arch,
            best_phase=best_phase,
            phase_score=phase_score,
            confidence=best_conf,
            phase_label=phase_label,
            all_states=dict(self._states),
        )

    def get_state(self) -> NarrativeResult:
        """Get current state without advancing."""
        return self._build_result()

    def reset(self) -> None:
        """Reset all archetype states to IDLE."""
        for state in self._states.values():
            state.phase = PHASE_IDLE
            state.phase_index = 0
            state.confidence = 0.0
            state.history.clear()
            state.trigger_log.clear()

File: sdk/java/src/NarrativeTracker.java
package com.hellohari;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

/**
 * Narrative State Machine for scam call phase tracking (Java port).
 *
 * Tracks the temporal progression of a scam call through manipulation phases:
 *   IDLE → HOOK → ESCALATE → ISOLATE → TRAP
 *
 * Loaded from the "narrative_phases" section of OSIF v2 patterns.json.
 */
public class NarrativeTracker {

    public static final String PHASE_IDLE = "IDLE";
    public static final String PHASE_HOOK = "HOOK";
    public static final String PHASE_ESCALATE = "ESCALATE";
    public static final String PHASE_ISOLATE = "ISOLATE";
    public static final String PHASE_TRAP = "TRAP";

    private static final Map<String, Integer> PHASE_SCORES = new HashMap<>();
    static {
        PHASE_SCORES.put(PHASE_IDLE, 0);
        PHASE_SCORES.put(PHASE_HOOK, 15);
        PHASE_SCORES.put(PHASE_ESCALATE, 40);
        PHASE_SCORES.put(PHASE_ISOLATE, 65);
        PHASE_SCORES.put(PHASE_TRAP, 95);
    }

    /** Per-archetype state. */
    private static class ArchetypeState {
        String archetypeId;
        String phase = PHASE_IDLE;
        int phaseIndex = 0;
        double confidence = 0.0;

        ArchetypeState(String id) { this.archetypeId = id; }
    }

    /** Phase definition from JSON. */
    private static class PhaseDef {
        String id;          // HOOK, ESCALATE, etc.
        String triggers;    // "context", "threat", "demand", "secrecy", etc.
        String description;
    }

    /** Archetype narrative config. */
    private static class ArchetypeNarrative {
        String label;
        List<PhaseDef> phases = new ArrayList<>();
        Map<String, Set<String>> extraKeywords = new HashMap<>(); // e.g. "secrecy_keywords"
    }

    private final Map<String, ArchetypeNarrative> narrativePhases = new HashMap<>();
    private final Map<String, ArchetypeState> states = new HashMap<>();

    // Archetype keyword data for trigger matching (from main patterns.json archetypes)
    private final Map<String, List<String>> archetypeContext = new HashMap<>();
    private final Map<String, List<String>> archetypeThreat = new HashMap<>();
    private final Map<String, List<String>> archetypeDemand = new HashMap<>();

    /**
     * Load narrative phases from the OSIF v2 JSON.
     *
     * @param root the full patterns.json root object
     */
    public void loadFromJson(JSONObject root) throws JSONException {
        // Load narrative_phases
        JSONObject npObj = root.optJSONObject("narrative_phases");
        if (npObj == null) return;

        Iterator<String> keys = npObj.keys();
        while (keys.hasNext()) {
            String archId = keys.next();
            JSONObject def = npObj.getJSONObject(archId);

            ArchetypeNarrative an = new ArchetypeNarrative();
            an.label = def.optString("label", "");

            JSONArray phasesArr = def.optJSONArray("phases");
            if (phasesArr != null) {
                for (int i = 0; i < phasesArr.length(); i++) {
                    JSONObject p = phasesArr.getJSONObject(i);
                    PhaseDef pd = new PhaseDef();
                    pd.id = p.getString("id");
                    pd.triggers = p.getString("triggers");
                    pd.description = p.optString("description", "");
                    an.phases.add(pd);
                }
            }

            // Load extra keyword sets (secrecy_keywords, urgency_keywords, etc.)
            Iterator<String> defKeys = def.keys();
            while (defKeys.hasNext()) {
                String key = defKeys.next();
                if (key.endsWith("_keywords")) {
                    JSONArray arr = def.optJSONArray(key);
                    if (arr != null) {
                        Set<String> kwSet = new HashSet<>();
                        for (int i = 0; i < arr.length(); i++) {
                            kwSet.add(arr.getString(i));
                        }
                        an.extraKeywords.put(key, kwSet);
                    }
                }
            }

            narrativePhases.put(archId, an);
            states.put(archId, new ArchetypeState(archId));
        }

        // Load archetype keyword lists for trigger matching
        JSONArray archetypes = root.optJSONArray("archetypes");
        if (archetypes != null) {
            for (int i = 0; i < archetypes.length(); i++) {
                JSONObject arch = archetypes.getJSONObject(i);
                String id = arch.getString("id");
                archetypeContext.put(id, jsonArrayToList(arch.optJSONArray("context")));
                archetypeThreat.put(id, jsonArrayToList(arch.optJSONArray("threat")));
                archetypeDemand.put(id, jsonArrayToList(arch.optJSONArray("demand")));
            }
        }
    }

    /**
     * Process a transcript chunk and advance narrative states.
     *
     * @param text transcript chunk
     * @return result with best phase info
     */
    public NarrativeResult advance(String text) {
        if (text == null || text.trim().isEmpty()) return buildResult();

        String lower = text.toLowerCase(Locale.ROOT);
        Set<String> words = new HashSet<>();
        for (String w : lower.split("\\s+")) {
            if (!w.isEmpty()) words.add(w);
        }

        for (Map.Entry<String, ArchetypeState> entry : states.entrySet()) {
            String archId = entry.getKey();
            ArchetypeState state = entry.getValue();
            ArchetypeNarrative narrative = narrativePhases.get(archId);
            if (narrative == null) continue;

            int idx = state.phaseIndex;
            if (idx >= narrative.phases.size()) continue;

            // Try to advance through multiple phases per chunk
            boolean advanced = false;
            while (idx < narrative.phases.size()) {
                PhaseDef nextPhase = narrative.phases.get(idx);
                String matched = checkTrigger(lower, words, archId, nextPhase.triggers, narrative);
                if (matched != null) {
                    state.phase = nextPhase.id;
                    state.phaseIndex = idx + 1;
                    state.confidence = Math.min(1.0, state.confidence + 0.25);
                    advanced = true;
                    idx = state.phaseIndex;
                } else {
                    break;
                }
            }

            if (!advanced) {
                // Phase skip: only after HOOK reached (phaseIndex >= 1)
                if (state.phaseIndex < 1) continue;

                for (int skipIdx = idx + 1; skipIdx < narrative.phases.size(); skipIdx++) {
                    PhaseDef skipPhase = narrative.phases.get(skipIdx);
                    // Use strict matching for skips (substring only, no stem)
                    String skipMatched = checkTriggerStrict(lower, archId, skipPhase.triggers, narrative);
                    if (skipMatched != null) {
                        int skipped = skipIdx - idx;
                        state.phase = skipPhase.id;
                        state.phaseIndex = skipIdx + 1;
                        state.confidence = Math.min(1.0, state.confidence + 0.25 - (0.05 * skipped));
                        break;
                    }
                }
            }
        }

        return buildResult();
    }

    /** Get current state without advancing. */
    public NarrativeResult getState() {
        return buildResult();
    }

    /** Reset all states. */
    public void reset() {
        for (ArchetypeState state : states.values()) {
            state.phase = PHASE_IDLE;
            state.phaseIndex = 0;
            state.confidence = 0.0;
        }
    }

    private String checkTrigger(String lower, Set<String> words, String archId,
                                String triggerType, ArchetypeNarrative narrative) {
        switch (triggerType) {
            case "context":
                return matchAny(lower, words, archetypeContext.getOrDefault(archId, List.of()));
            case "threat":
                return matchAny(lower, words, archetypeThreat.getOrDefault(archId, List.of()));
            case "demand":
                return matchAny(lower, words, archetypeDemand.getOrDefault(archId, List.of()));
            default:
                String extraKey = triggerType + "_keywords";
                Set<String> extra = narrative.extraKeywords.getOrDefault(extraKey, Set.of());
                return matchAny(lower, words, new ArrayList<>(extra));
        }
    }

    /** Strict trigger check for phase skipping — substring only, no stem matching. */
    private String checkTriggerStrict(String lower, String archId,
                                      String triggerType, ArchetypeNarrative narrative) {
        switch (triggerType) {
            case "context":
                return matchAnyStrict(lower, archetypeContext.getOrDefault(archId, List.of()));
            case "threat":
                return matchAnyStrict(lower, archetypeThreat.getOrDefault(archId, List.of()));
            case "demand":
                return matchAnyStrict(lower, archetypeDemand.getOrDefault(archId, List.of()));
            default:
                String extraKey = triggerType + "_keywords";
                Set<String> extra = narrative.extraKeywords.getOrDefault(extraKey, Set.of());
                return matchAnyStrict(lower, new ArrayList<>(extra));
        }
    }

    private String matchAny(String lower, Set<String> words, List<String> keywords) {
        for (String kw : keywords) {
            String kwLower = kw.toLowerCase(Locale.ROOT);
            if (lower.contains(kwLower)) return kw;
            if (kwLower.length() >= 4) {
                for (String word : words) {
                    if (word.length() >= 4) {
                        if (word.startsWith(kwLower) || kwLower.startsWith(word)) return kw;
                    }
                }
            }
        }
        return null;
    }

    /** Strict substring-only match — no stem/prefix matching. */
    private String matchAnyStrict(String lower, List<String> keywords) {
        for (String kw : keywords) {
            if (lower.contains(kw.toLowerCase(Locale.ROOT))) return kw;
        }
        return null;
    }

    private NarrativeResult buildResult() {
        String bestArch = null;
        String bestPhase = PHASE_IDLE;
        int bestIdx = 0;

        for (Map.Entry<String, ArchetypeState> entry : states.entrySet()) {
            ArchetypeState state = entry.getValue();
            if (state.phaseIndex > bestIdx) {
                bestIdx = state.phaseIndex;
                bestArch = entry.getKey();
                bestPhase = state.phase;
            }
        }

        int phaseScore = PHASE_SCORES.getOrDefault(bestPhase, 0);
        String phaseLabel = "";
        if (bestArch != null) {
            ArchetypeNarrative narrative = narrativePhases.get(bestArch);
            if (narrative != null) {
                for (PhaseDef p : narrative.phases) {
                    if (p.id.equals(bestPhase)) {
                        phaseLabel = p.description;
                        break;
                    }
                }
            }
        }

        return new NarrativeResult(bestArch, bestPhase, phaseScore, phaseLabel);
    }

    private List<String> jsonArrayToList(JSONArray arr) {
        List<String> list = new ArrayList<>();
        if (arr == null) return list;
        for (int i = 0; i < arr.length(); i++) {
            try { list.add(arr.getString(i)); } catch (JSONException ignored) {}
        }
        return list;
    }

    /** Result of narrative analysis. */
    public static class NarrativeResult {
        public final String bestArchetype;
        public final String phase;
        public final int phaseScore;
        public final String phaseLabel;

        NarrativeResult(String bestArchetype, String phase, int phaseScore, String phaseLabel) {
            this.bestArchetype = bestArchetype;
            this.phase = phase;
            this.phaseScore = phaseScore;
            this.phaseLabel = phaseLabel;
        }
    }
}
